### DISCUSSION TITLE
Implementing generic ble-peripheral traits

### DISCUSSION THREAD

**Discussion by jasta** [bluez/bluer#109] [Ideas]
Hi all, I've been working on the [Rust Matter SDK](https://github.com/project-chip/rs-matter) (the new hotness in connected home protocols) and was looking for a generic path to support embedded BLE peripheral support and realized the ecosystem had a lot missing to do this in a well scoped and straight forward way.  Specifically, there is no set of traits to implement BLE peripheral support that a library like `rs-matter` can consume.  Enter `ble-peripheral`, where I designed these traits to be embedded and desktop friendly to make it a really great experience to test on, e.g., a  Linux host while ultimately targeting a peripheral running on firmware.

I started with a reference implementation for BlueR and would love some feedback: https://github.com/jasta/bluer-ble-peripheral

Ultimately my goal would be to get this code adopted by the bluez github account (or maybe just made into a feature in the bluer crate) to ensure quality support from this great community.

Oh and one small thing, I couldn't figure out a good way to detect connect/disconnect events and map them into the ble-peripheral traits.  Did I miss something or is that intentionally not supported (and if so, why)?

Thanks for the great project all, looking forward to feedback!

**Comment by surban** [comment]
Hi!

It's definitely a really cool idea to abstract the GATT server interface as traits.

Adding this to bluer is probably out-of-scope though, because we strictly follow the features provided by BlueZ. However, I don't see this as a problem, since you can publish an external adapter crate which provides the implementations of your traits using bluer.

> Oh and one small thing, I couldn't figure out a good way to detect connect/disconnect events and map them into the ble-peripheral traits. Did I miss something or is that intentionally not supported (and if so, why)?

I have not tested this myself, but have you tried receiving changes to the [Connected property](https://docs.rs/bluer/latest/bluer/enum.DeviceProperty.html#variant.Connected) using [`Device::events`](https://docs.rs/bluer/latest/bluer/struct.Device.html#method.events)? When you are a GATT server and a client connects, it should appear as device that is [enumerable](https://docs.rs/bluer/latest/bluer/struct.Adapter.html#method.device_addresses).

Keep up the great work, and I'm excited to see how your project evolves. Please keep the community posted on your progress and any support you might need along the way!
